ec2_asg and ec2_asg_facts module improvements

willthames opened the pull request:
##### SUMMARY
Return target group information for both ec2_asg and ec2_asg_facts
modules

Provide RETURN documentation for ec2_asg module

PEP8 fixes for ec2_asg_facts

##### ISSUE TYPE
 - Feature Pull Request

##### COMPONENT NAME
ec2_asg and ec2_asg_facts

##### ANSIBLE VERSION
```
ansible 2.4.0 (devel fd5db52fd9) last updated 2017/05/31 00:18:27 (GMT +1000)
  config file = 
  configured module search path = [u'/Users/will/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /Users/will/src/ansible/lib/ansible
  executable location = /Users/will/src/ansible/bin/ansible
  python version = 2.7.13 (default, Dec 18 2016, 07:03:39) [GCC 4.2.1 Compatible Apple LLVM 8.0.0 (clang-800.0.42.1)]
```


##### ADDITIONAL INFORMATION
Unfortunately there remains inconsistencies between ec2_asg and ec2_asg_facts return data.

In an ideal world, we'd used the better data, but ec2_asg is marked as stable interface and ec2_asg_facts is marked as preview - but ec2_asg_facts has the better data.

It's now documented, at least.

Commit:
ec2_asg and ec2_asg_facts module improvements

Return target group information for both ec2_asg and ec2_asg_facts
modules

Provide RETURN documentation for ec2_asg module

PEP8 fixes for ec2_asg_facts

willthames commented:
cc @wimnat

s-hertel commented on lib/ansible/modules/cloud/amazon/ec2_asg.py:
```diff
@@ -1023,16 +1165,15 @@ def wait_for_term_inst(connection, module, term_instances):
     batch_size = module.params.get('replace_batch_size')
```
If you're getting rid of unused variables you could get rid of this line.

willthames replied:
Done, not sure why I hadn't spotted that

s-hertel commented on lib/ansible/modules/cloud/amazon/ec2_asg_facts.py:
```diff
@@ -265,6 +294,7 @@ def find_asgs(conn, module, name=None, tags=None):
                     }
                 ],
+                "launch_config_name": "public-webapp-production-1",
                 "launch_configuration_name": "public-webapp-production-1",
```
Did you add launch_config_name in addition to launch_configuration_name to the return just to be more consistent with ec2_asg?

willthames replied:
Yes, that's exactly it. Not sure it's needed tbh, there are so many other inconsistencies.

s-hertel commented on lib/ansible/modules/cloud/amazon/ec2_asg_facts.py:
```diff
@@ -328,7 +368,17 @@ def find_asgs(conn, module, name=None, tags=None):
+            # compatibility with ec2_asg module
+            asg['launch_config_name'] = asg['launch_configuration_name']
+            # workaround for https://github.com/ansible/ansible/pull/25015
+            if 'target_group_ar_ns' in asg:
```
:/

willthames replied:
#25015 is an unmerged fix for this

s-hertel reviewed (commented):
Looks good.

ryansb commented on lib/ansible/modules/cloud/amazon/ec2_asg.py:
```diff
@@ -320,6 +450,19 @@ def get_properties(autoscaling_group):
+                                      region=region,
+                                      endpoint=ec2_url,
+                                      **aws_connect_params)
+        target_groups = elbv2_connection.describe_target_groups(TargetGroupArns=properties['target_group_arns'])['TargetGroups']
```
Please add in an `AWSRetry.backoff()` here in case the describe gets rate limited. Also, you may want to add a paginator in case there are lots of target groups. I've seen 100+ in accounts that use ECS heavily.

willthames replied:
That many target groups in a single ALB? I guess it's possible.

willthames replied:
Are AWSRetry and pagination both required? We have no current examples of that, although guess it would be 
```
AWSRetry(tries=5, delay=5, backoff=2.0))(tg_paginator.paginate)(TargetGroupArns=asg['target_group_arns'])
```

Commit:
ec2_asg: use pagination when describing target groups

In case an ASG has 100s of target groups, ensure that
we get the full result using build_full_result

willthames commented:
bot_broken

ryansb commented:
shipit
